src/nunavut/lang/c/templates/_deserialize.j2

```diff
@@ -0,0 +1,170 @@
+    {{ _deserialize_array(type, name, ref, offset) }}
+    {%- elif type is CompositeType -%}
+    {%- if name == '' -%}
+        {%- set ref_name = ref -%}
+    {%- else -%}
+        {%- set ref_name = ref + (name | id) -%}
+    {%- endif -%}
+    {%- if offset.is_aligned_at_byte() -%}
+    {#- Data is byte aligned: use generated function -#}
+    rc = {{ type | full_reference_name }}_deserialize(&({{ ref_name }}), offset, in_buffer, buf_size_bytes);
+    if (rc < 0)
+    {
+        return rc;
+    }
+    offset = rc;
+    {%- else -%}
+    {#- Not byte-aligned: inline serialization -#}
+    {{ _deserialize_composite(type, ref_name + '.', offset) }}
+    {%- endif -%}
+    {%- endif -%}
+{%- endmacro %}
+
+{%- macro _deserialize_composite(type, ref, offset) -%}
+    {%- if type is StructureType -%}
+    {%- for f, offset in type.iterate_fields_with_offsets(offset) %}
+    // Begin Structure: {{ f.data_type }}
+    {{ _deserialize_any(f.data_type, f.name, ref, offset) }}
+    // End Structure: {{ f.data_type }}
+    {%- endfor -%}
+    {%- elif type is UnionType %}
+    // Begin Union: {{ type }}
+    {{ _deserialize_int(type.tag_field_type, '_tag_', ref) }}
+    offset += {{ type.tag_field_type.bit_length }};
+    {%- for f, offset in type.iterate_fields_with_offsets(offset) -%}
+    {%- if loop.first %}
+    if (
+    {%- else %}
+    else if (
+    {%- endif -%}
+    {%- if ref.endswith('.') -%}
+        {{ type | full_reference_name }}_is_{{ f.name }}(&({{ ref[:-1] }})))
+    {%- elif ref.endswith('->') -%}
+        {{ type | full_reference_name }}_is_{{ f.name }}({{ ref[:-2] }}))
+    {%- endif %}
+    {
+        {{ _deserialize_any(f.data_type, f.name, ref, offset) | indent(width=4) }}
+    }
+    {%- endfor %}
+    else
+    {
+        return -NUNAVUT_ERR_INVALID_TAG;
+    }
+    // End Union: {{ type }}
+    {%- endif %}
+{%- endmacro %}
+
+{#
+    Array serialization, handles fixed or variable length
+#}
+{%- macro _deserialize_array(type, name, ref, base_offset) %}
+    {#- wrangle length value if variable length array #}
+    {%- if type is VariableLengthArrayType -%}
+    {{ _deserialize_int(type.length_field_type, name + "_length", ref) }};
+    offset += {{ type.length_field_type.bit_length }};
+    if ({{ ref }}{{ name | id }}_length > {{ type.capacity }})
+    {
+        return -NUNAVUT_ERR_INVALID_LEN;
+    }
+    {%- endif %}
+    {%- set loop_type = type.element_type %}
+    {%- set loop_ref = ref + (name | id) + "[i]" %}
+    {%- if loop_type is ArithmeticType and loop_type.standard_bit_length %}
+    {#- There are some options for memcpy if it's a standard data type #}
+    {%- if base_offset.is_aligned_at_byte() and option_target_endianness == "little" %}
+    {#- Data is byte-aligned, so we can use memcpy #}
+    {%- if type is VariableLengthArrayType -%}
+    memcpy((void*)out_instance->{{ name }}, (void*)((uint8_t*)in_buffer + (offset >> 3)), out_instance->{{ name }}_length * {{ (loop_type.bit_length / 8) | int }});
```

Comment:
You don't check `buf_size_bytes` before using `in_buffer` here. Please do and return appropriate error (see comment in serialization.j2)